### PR TITLE
Tunable madvise for file storage

### PR DESCRIPTION
**Pull request by gquintard**

### CODE REVIEW

**Comment by gquintard** [comment]
Following up on Nils comment: DONTDUMP seems to exist only on Linux.

**Comment by nigoroll** [comment]
yes, would need to be auto-crapped

File: bin/varnishd/storage/storage_file.c [hunk 1]
```diff
@@ -129,6 +131,16 @@ smf_init(struct stevedore *parent, int ac, char * const *av)
...
 	}
+	if (ac > 3) {
+		if (!strcmp(av[3], "normal"))
+			advice = MADV_NORMAL;
+		else if (!strcmp(av[3], "random"))
+			advice = MADV_RANDOM;
+		else if (!strcmp(av[3], "sequential"))
+			advice = MADV_SEQUENTIAL;
+		else
+			ARGV_ERR("(-s file) invalid advice: \"%s\"");
```
Comment (fgsch) [inline bin/varnishd/storage/storage_file.c]: Missing argument.

File: doc/sphinx/users-guide/storage-backends.rst [hunk 2]
```diff
@@ -94,6 +94,13 @@ have many small objects.
 File performance is typically limited to the write speed of the
 device, and depending on use, the seek time.
 
+'advice' dictates what Varnish tells the system to optimize reads. Depending
+on your OS, disks and object sizes, it can be beneficial yo tweak this. The
```
Comment (fgsch) [inline doc/sphinx/users-guide/storage-backends.rst]: yo -> to. This sentence could be improved.

File: bin/varnishd/storage/storage_file.c [hunk 3]
```diff
@@ -129,6 +131,17 @@ smf_init(struct stevedore *parent, int ac, char * const *av)
...
+	if (ac > 3) {
+		if (!strcmp(av[3], "normal"))
+			advice = MADV_NORMAL;
+		else if (!strcmp(av[3], "random"))
+			advice = MADV_RANDOM;
+		else if (!strcmp(av[3], "sequential"))
+			advice = MADV_SEQUENTIAL;
+		else
+			ARGV_ERR("(-s file) invalid advice: \"%s\"",
+					strcmp(av[3]);
```
Comment (fgsch) [inline bin/varnishd/storage/storage_file.c]: pasto?

Comment (gquintard) [inline bin/varnishd/storage/storage_file.c]: yeah, I really shouldn't do these things hastily...

Thanks

**Comment by nigoroll** [comment]
irc transcript:

> (13:30:36) scn: 1997 - is it really necessary to push that decision to the user?
> (13:30:53) phk: I have no operational input/experience on that...
> (13:31:45) fgs: I thought the reason is that it makes sense under some scenarios
> (13:31:50) fgs: not necessarily all
> (13:32:19) slink: actually the only thing I'd have an application for would be MADV_DONTDUMP. Remember what dho said on Vcon? We can't dump cores - our caches are too big. So for no matter what stv there should be the option not to dump it. For file, this is comparably cheap, for malloc we'd need to put it into jemalloc probably
> (13:32:35) scn: i think martin looked at this for mse(|2). only MADV_RANDOM is in use there.
> (13:32:44) slink: in general I think having auto tuning is better than manual tuning is better than no tuning.
> (13:32:47) scn: seq in -sfile for historic reasons?
> (13:33:32) slink: guiallaume not present? Does anyone know his motivation?
> (13:34:10) dridi_: slink: guillaume is enjoying the us independence day today
> (13:34:15) phk: If you have _really_ big objects, MADV_NORMAL *might* be better than MADV_RANDOM, but it takes a number of lucky circumstances
> (13:34:24) slink: ok so should we talk about this when he's present?
> (13:34:31) phk: yeah, think so

**Comment by bsdphk** [comment]
OK

**Comment by hermunn** [comment]
Backport review: This is new functionality that will not be ported back to 4.1.

**Pull request merged**